File: leetcode/844. Backspace String Compare/844.js
// https://helloacm.com/the-backspace-string-compare-algorithm/
// https://leetcode.com/problems/backspace-string-compare/

/**
 * @param {string} S
 * @param {string} T
 * @return {boolean}
 */
var backspaceCompare = function(S, T) {
    const build = (S) => {
        let st = [];        
        for (let i = 0, len = S.length; i < len; ++ i) {
            if (S[i] == '#') {
                st.pop();
            } else {
                st.push(S[i]);
            }
        }
        return st.join('');
    }
    return build(S) === build(T);
};

File: leetcode/844. Backspace String Compare/844.cpp
// https://helloacm.com/the-backspace-string-compare-algorithm/
// https://leetcode.com/problems/backspace-string-compare/

class Solution {
public:
    bool backspaceCompare(string S, string T) {
        stack<char> st1 = realString(S);
        stack<char> st2 = realString(T);
        if (st1.size() != st2.size()) return false;
        while (st1.size() > 0) {
            char p1 = st1.top();
            char p2 = st2.top();
            st1.pop();
            st2.pop();
            if (p1 != p2) return false;
        }
        return true;
    }
    
private:
    stack<char> realString(string S) {
        stack<char> st;    
        for (int i = 0; i < S.size(); ++ i) {
            if (S[i] == '#') {
                if (st.size() > 0) {
                    st.pop();
                }
            } else {
                st.push(S[i]);
            }
        }
        return st;
    }
};
